### PR TITLE
Accelerate exact Gaussian-hidden pseudolikelihood

### PR DESCRIPTION
**Pull request by cossio**
## Summary

- evaluate the quadratic Gaussian hidden cumulant analytically for exact Binary, Spin, Potts, and PottsGumbel pseudolikelihood
- retain multidimensional batches, unbatched inputs, dense Potts inputs, signed `γ` semantics, Float32/Float64 and mixed-precision behavior, integer parameters, and generic array backends
- scale only one Potts site's weights at a time, avoiding a second full weight-sized temporary
- reuse hidden-input buffers in the generic non-Gaussian exact paths
- form Potts relative logits before adding visible-field differences, avoiding precision loss under large common hidden offsets

## Evaluation and design

For Gaussian hidden units,

```text
Γ(z + u) - Γ(z) = u' (z / |γ|) + 1/2 u' (u / |γ|).
```

The retained implementation shares the hidden mean and uses matrix products for the visible conditional fields. Binary and Spin remain fully vectorized. A fully streaming Binary formulation reduced allocation only modestly but was 10–29× slower at moderate/large sizes, so it was rejected. Potts keeps the global field product but builds the scaled weights and Gram matrix one site at a time. Relative hidden, visible-field, and dense-input terms are accumulated separately so small visible-field differences are not added to a large common hidden contribution.

The generic buffer-reuse edits are independently useful: across Binary-Binary and Potts-Binary cases they reduce allocated bytes by 42–50% (1.73–1.99×) with essentially neutral to slightly improved runtime (1.00–1.01×). Cumulant differences are formed before visible-field terms, preserving the original numerical ordering while retaining buffer reuse.

## Benchmarks

Methodology: Julia 1.12.6 on `znver5`, one Julia thread and one BLAS thread, deterministic warmed inputs, adaptive repeated calls, median wall time, and `Base.@allocated`. `old` recreates the exact specialized loops from `origin/master`; `fast` is commit `5641606b`. Float32 is used unless noted.

| workload | old time | fast time | speedup | old alloc | fast alloc | alloc reduction |
|---|---:|---:|---:|---:|---:|---:|
| Binary N=784, M=256, B=256 | 285.9 ms | 5.07 ms | 56.4× | 416.0 MB | 4.54 MB | 91.5× |
| Spin N=256, M=128, B=128 | 25.2 ms | 0.675 ms | 37.3× | 34.5 MB | 0.790 MB | 43.7× |
| Potts q=21, N=100, M=256, B=64 | 192.5 ms | 2.34 ms | 82.2× | 285.0 MB | 3.85 MB | 74.0× |
| dense Potts Float64 q=5, N=64, M=128, B=64 | 20.9 ms | 0.503 ms | 41.6× | 46.9 MB | 1.25 MB | 37.4× |

No runtime crossover regression appeared in the scans:

- Binary Float32/Float64, N=1…256, M=64, B=1/32: 1.7–24.0× faster
- Potts Float32, q=3 with N=1…128 and q=21 with N=1…32, M=64, B=1/32: 1.7–27.5× faster
- against the generic substitution-matrix reference on small cases: 2.9–16.0× faster with 2.1–13.6× lower allocation

## Correctness and numerical stability

The optimized results were compared directly with the generic substitution-matrix implementation for Binary, Spin, Potts, and PottsGumbel visible layers with Gaussian hidden units across:

- Float32 and Float64
- multidimensional `(2, 3)` batches and unbatched inputs
- one-hot and dense Potts inputs
- multidimensional hidden layers
- nontrivial positive and negative `γ`
- high-curvature signed-`γ` stress cases
- Float64 visible fields with Float32 weights/hidden parameters
- integer weights and Gaussian parameters with fractional inverse precision
- JLArrays with scalar indexing disabled

Maximum scaled differences were below `4e-7` for ordinary Float32 cases and below `8e-16` for Float64. A stress case with `|γ| = 1e-3` remained finite and agreed within `2.3e-7` scaled error in Float32.

Review found three edge cases in the initial PR revision, all fixed in `5641606b`:

1. Float32 Potts fields `[0, 1, -1]` under a `1e8` common hidden contribution were rounded away, returning `-log(3)` instead of `-1.4076059`. Relative hidden terms are now formed before visible fields.
2. Float64 visible fields could be downcast when weights and hidden parameters were Float32. The Gaussian calculation type now includes visible parameters.
3. Integer Potts weights with `γ = 2` attempted to store `1/2`-scaled weights in an integer buffer. Gaussian calculations now use a floating promoted type and allocate temporaries from the computed linear result.

Regression coverage includes Binary, Spin, Potts, and PottsGumbel mixed precision, one-hot and dense large-offset Potts inputs, integer Potts parameters, and CPU/JLArray backends.

## Validation

- `julia --project=test test/pseudolikelihood.jl`
- `julia --project=test test/jlarrays.jl`
- extended reference matrix: 72 CPU assertions, 96 JLArrays assertions, and 12 signed-`γ` stress assertions
- `julia --project=. -e 'using Pkg; Pkg.test()'`

All passed on Julia 1.12.6. The branch is rebased onto current `origin/master` (`e77b1b24`).

### CODE REVIEW

**Review by cossio — COMMENTED** [review]
No blocking issues found.

I verified the core identity by hand against this codebase's Gaussian layer (`src/layers/gaussian.jl`): `cgfs` uses `gauss_cfg(θ,γ) = θ²/|2γ| - log(|γ|/2π)/2`, so for a single hidden unit with total field `z`, `Γ(z+u) - Γ(z) = u·z/|γ| + u²/(2|γ|)`, matching the PR's stated identity and its use of `mean_from_inputs`-style `z/|γ|` (`hidden_mean` in `_gaussian_hidden_pseudolikelihood_context`, `src/pseudolikelihood.jl`).

For the Binary/Spin fast path (`_log_pseudolikelihood_exact_2states(rbm::RBM{<:Union{Binary,Spin},<:Gaussian}, ...)`), substituting `u_m = δ·w_{jm}` into the identity reproduces `minusΔF = (θv + linear)·δ + quadratic·δ²` exactly, which equals the generic `θ·δ + (Γ1-Γ0)` used in the non-specialized 2-states path and in `log1pexp`.

For the Potts/Gaussian fast path (`log_pseudolikelihood_exact(rbm::RBM{<:Potts,<:Gaussian}, ...)`), I expanded `ΔΓ(x) = Γ(z+u)-Γ(z)` with `u = w_x - w_c` (`w_c` being the current, possibly dense, site contribution) and confirmed the `current` term (`θ_current + linear_current - w_c'(w_c/|γ|)/2`) subtracted from `logits` exactly cancels to reproduce `-ΔF(x)` and, critically, forces `logits[current_state] = 0` before `logsumexp` — this is the stability fix described in the PR body (WIP returning `0` instead of `-log(3)`), and it's covered by the new regression test with `θ=0, γ=-1, θ_hidden=1e8`.

Both fast paths are cross-checked in the diff against the generic substitution-matrix reference (`invoke(log_pseudolikelihood_exact, Tuple{RBM,AbstractArray}, ...)`) across Float32/Float64, multidimensional batches, unbatched input, dense vs one-hot Potts, and PottsGumbel, plus a dedicated JLArray/`allowscalar(false)` testset in `test/jlarrays.jl` exercising multidimensional hidden layers and signed `γ`. CI is green across Julia 1.10/1/downgrade-compat. The CHANGELOG `## Unreleased` entry accurately describes the behavior-preserving performance change, and `Project.toml` keeps its `-DEV` suffix.

No issues, no nits.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/pseudolikelihood.jl [hunk 1]
```diff
@@ -328,11 +379,57 @@ function log_pseudolikelihood_exact(rbm::RBM{<:Potts}, v::AbstractArray)
...
+        # Keep the x-independent term explicitly so this remains correct for
+        # dense Potts inputs as well as the usual one-hot representation.
+        current = vec(sum(
+            Vsite .* (
+                reshape(θsite, q, 1) .+ linear_site .- gram_v ./ 2
+            );
+            dims=1,
+        ))
+        logits .= reshape(θsite, q, 1) .+ linear_site .+
+            reshape(LinearAlgebra.diag(gram), q, 1) ./ 2 .- gram_v
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/pseudolikelihood.jl]: Fixed in `5641606b`. The Potts path now subtracts the current hidden contribution before adding the relative visible-field term, with the Gram/dense-input correction accumulated separately. The reported Float32 case with fields `[0, 1, -1]` and a `1e8` common hidden offset now returns `-1.4076059`; regressions cover one-hot and dense inputs on CPU and the large-offset path on JLArrays. I also restored the same stable ordering in the generic buffer-reuse paths by forming `Γx - Γ0` before adding field differences.

File: src/pseudolikelihood.jl [hunk 2]
```diff
@@ -284,24 +284,74 @@ function log_pseudolikelihood_sites(
...
+    B = prod(batch_sz)
+    vflat = reshape(v, length(rbm.visible), B)
+
+    # The Gaussian cgf is quadratic. Its change under a hidden-input shift u is
+    #
+    #   Γ(z + u) - Γ(z) = u' * (z / |γ|) + u' * (u / |γ|) / 2.
+    #
+    # Precompute the terms shared by all visible-site substitutions. This turns
+    # the repeated hidden cgf evaluations below into matrix multiplications.
+    T = promote_type(eltype(rbm.w), eltype(rbm.hidden.par))
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/pseudolikelihood.jl]: Fixed in `5641606b`. The Gaussian calculation type is now `float(promote_type(eltype(visible.par), eltype(hidden.par), eltype(w)))`, so Float64 visible fields are no longer downcast when weights/hidden parameters are Float32. Regression coverage includes Binary, Spin, Potts, and PottsGumbel mixed-precision models; Potts is also exercised on JLArrays and returns Float64 matching the generic reference.

File: src/pseudolikelihood.jl [hunk 3]
```diff
@@ -328,11 +379,57 @@ function log_pseudolikelihood_exact(rbm::RBM{<:Potts}, v::AbstractArray)
...
+    @assert size(rbm.visible) == size(v)[1:ndims(rbm.visible)]
+    batch_sz, B, vflat, wflat, invγ, hidden_mean =
+        _gaussian_hidden_pseudolikelihood_context(rbm, v)
+    q = colors(rbm.visible)
+    nsites = prod(sitesize(rbm.visible))
+    linear = wflat * hidden_mean
+    θflat = convert_eltype(eltype(linear), vec(rbm.visible.θ))
+    lPL = zero(similar(linear, B))
+    logits = similar(linear, q, B)
+    wscaled_site = similar(wflat, q, length(rbm.hidden))
```
[the remark this answers is not in the document]
Comment (cossio) [reply src/pseudolikelihood.jl]: Fixed in `5641606b`. Gaussian calculations now promote integer parameter models to a floating type, and `wscaled_site` is allocated from the computed `linear` array rather than raw weights. The added integer Potts regression (`w::Array{Int}`, `γ = 2`) returns Float64 and matches the generic exact path without `InexactError`.

**Review by cossio — COMMENTED** [review]
Reviewed the new commit (`5641606`, "Fix Gaussian pseudolikelihood edge cases") added since my last pass. No blocking issues found.

**Precision fix in `log_pseudolikelihood_exact(rbm::RBM{<:Potts,<:Gaussian}, ...)`** (`src/pseudolikelihood.jl`): splitting the "current-state" correction into three separate subtractions (linear term, θ term, Gram term) before combining is mathematically equivalent to the prior single combined `current` subtraction — the additive per-batch constant removed from `logits` is identical in both forms (I re-derived it by hand). The benefit is purely numerical: it avoids adding a small term (`linear_site`/`gram_v`, O(1)) to a large common offset (`θsite[c] ~ 1e8`) before the cancelling subtraction, which is exactly the precision-loss pattern that produced the earlier wrong result (`0` instead of `-log(3)`, per this PR's description). New regression tests confirm this: `test/pseudolikelihood.jl` now checks exact expected values against an asymmetric `θ_offset = [0, 1, -1]` Potts field under a `1e8` Gaussian hidden offset (rather than only `-log(q)` for a symmetric field), plus mixed-precision and integer-parameter cases; `test/jlarrays.jl` mirrors two of these on `JLArray` with `allowscalar(false)`.

**Widened `T` in `_gaussian_hidden_pseudolikelihood_context`**, now `float(promote_type(eltype(rbm.visible.par), eltype(rbm.hidden.par), eltype(rbm.w)))` (previously omitted the visible layer's eltype): correct. It preserves a `Float64` visible field's precision when hidden/weights are `Float32` (new tests assert `eltype(result) == Float64`), and `float(...)` additionally makes an all-`Int` RBM viable since `invγ = inv.(abs.(γh))` requires a floating type — covered by the new integer-parameter test.

Bundled `AGENTS.md`/`CLAUDE.md` changes are unrelated repo-instruction sync (matches current `CLAUDE.md` on `master`), not part of the fix itself.

CI was still running the Julia test matrix on this commit at review time (CodeQL/Analyze/claude-review already green); worth confirming those finish green before merge, but I see no correctness concern in the diff.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by cossio — COMMENTED** [review]
Reviewed the analytic Gaussian-hidden pseudolikelihood fast paths.

I re-derived the math independently: for a Gaussian hidden layer, `cgf(z) = z²/(2|γ|) + const`, so `Γ(z+u) - Γ(z) = u'(z/|γ|) + u'(u/|γ|)/2` as stated. Checked this against:
- The Binary/Spin path (`_log_pseudolikelihood_exact_2states` Gaussian specialization): `linear = W·(z/|γ|)` and `quadratic = diag(W diag(1/|γ|) W')/2`, combined with `δ` and `δ²` (using `flip=1` for Binary, `flip=2` for Spin) — matches.
- The Potts path (`log_pseudolikelihood_exact(rbm::RBM{<:Potts,<:Gaussian}, ...)`): expanding `u = w_x - Wold` in the quadratic and linear terms reproduces exactly the `linear_site`/`gram`/`gram_v`/`current` terms in the diff, including the x-independent pieces that cancel under `logsumexp`. The dense-Potts-input case (non-one-hot `Vsite`) doesn't break this since the derivation never assumed a one-hot indicator.
- Confirmed `T = float(promote_type(visible.par, hidden.par, w))` includes the visible dtype (fixes the earlier Float64-downcast issue) and that `wflat`/`hidden_mean` are promoted before use (fixes the earlier integer-buffer issue).
- `LinearAlgebra.diag(gram)` is implemented as a single fancy-index `A[diagind(A)]` call, not a scalar-indexing loop, so it's fine under `JLArrays.allowscalar(false)`.

No correctness issues found. Nice, thorough validation section (mixed precision, integer params, large common hidden offsets, JLArrays). No blocking issues.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by cossio — COMMENTED** [review]
## Review

Given this touches core numerical correctness (`log_pseudolikelihood`), I checked out the branch, ran the existing test suite, and independently re-derived the analytic Gaussian-cumulant math by hand rather than trusting the description. I also wrote adversarial tests beyond what's in the PR: |γ|=1e-3 stress, γ=0, integer visible/hidden parameters, empty batch (B=0), unbatched input, and a 3D Potts site shape `(3,2,2)` with multi-dim hidden `(2,3)` and 3D batch — all compared against the generic reference implementation.

**Correctness — verified, no bugs found:**
- Hand-derived `Γ(z+u)-Γ(z) = u'(z/|γ|) + 1/2 u'(u/|γ|)` and confirmed the Binary/Spin fast path (`src/pseudolikelihood.jl:333-356`) matches exactly, including the `abs2(flip)` vs `abs2(δ)` simplification (valid since `|δ|=|flip|` regardless of sign).
- Full algebraic decomposition of the Potts fast path (`src/pseudolikelihood.jl:396-440`) into six terms confirms it exactly reconstructs `-ΔF_x` — the claimed reordering fix (relative-hidden terms before visible fields) is genuinely implemented, confirmed by the passing `1e8`-offset tests in both `test/pseudolikelihood.jl:213-231` and `test/jlarrays.jl:213-230`.
- All three "bugs fixed during review" (Float32 precision loss under large common offset, Float64 downcast with Float32 weights, integer Potts weights breaking under γ scaling) are genuinely fixed in the code — not just described. `T = float(promote_type(...))` correctly includes `visible.par`; buffers are allocated via `similar(linear, ...)` at the promoted floating type, not the original integer eltype.
- Multidimensional/batch shape handling verified correct beyond the PR's own tests: my independent 3D Potts/multi-dim-hidden/3D-batch test matched the generic reference to ~1e-6 (float32).
- No scalar indexing introduced — `test/jlarrays.jl`'s new testset (21 assertions) passes under `allowscalar(false)`, including the `diag(gram)` call in the Potts path.
- Edge cases (γ=0 → NaN in both fast path and reference, consistent with pre-existing degenerate-layer behavior; empty batch; unbatched input) behave identically to the reference.
- `CHANGELOG.md` entry present and appropriate for this user-facing performance change.

**Non-blocking suggestions:**
- The PR description cites "72 CPU assertions, 96 JLArrays assertions, 12 signed-γ stress assertions" and a specific `|γ|=1e-3` result, but the actual committed diffs to `test/pseudolikelihood.jl`/`test/jlarrays.jl` show 29 and 21 new assertions respectively, and I couldn't locate a checked-in `|γ|=1e-3` case. My own independently-written version of that stress test passed, so this isn't a correctness problem, but the specific numbers in the description aren't reproducible from the committed test suite alone — worth tightening if those figures are meant to be verifiable by future readers.
- `gram`/`gram_v` in the Potts path (`src/pseudolikelihood.jl:416-417`) are freshly allocated each `j` iteration rather than reusing a preallocated buffer like `wscaled_site`/`current`/`logits` do — minor missed allocation-reduction, dwarfed by the overall speedup.
- The integer-Potts-parameter regression test is CPU-only; `test/jlarrays.jl` has no equivalent integer-parameter case, so bug fix #3 isn't cross-validated on the GPU-array-typed code path specifically.

**Approve** — the math checks out by hand-derivation, all three claimed bug fixes are genuinely present, and independent adversarial testing beyond the PR's own coverage found nothing wrong.

---
_Generated by [Claude Code](https://claude.ai/code)_
